### DISCUSSION TITLE
registry-image-resource unpack layers that had residual files from older layers

### DISCUSSION THREAD

**Discussion by itsai220** [concourse/concourse#6644] [Help & Support]
Hi guys,

This is something I noticed when building one of our service.

we built a build image from a base image that we needed to clean the go version properly and install newer go version. When inspecting the image we made sure everything looks correct with the correct go version. Our devs are able to run this image locally and build the service.

However, when using concourse and using registry-image-resource to fetch the build image, the build breaks because the files that was installed in the base image still exist, the registry-image-resource seems to unpack the older layer, or kept the older files even though it was cleaned up in the newer image layer.

To go workaround this i had to using docker-image-resource which is a simple docker pull and docker export to export the content to the image of the top layer.

in detail:
```
base image has go version 1.10 installed which has files under usr/local/go/src/runtime/internal/atomic/atomic_amd64x.go
in our build image dockerfile we run clean on the old go version, and installed new go version 1.14 which at this point go has moved atomic_amd64x.go to atomic_amd64.go

when inspecting the image everything is cleaned, however when registry-image-resource pulls,both atomic_amd64x.go and atomic_amd64.go are present and thus build fails.

there are other old files around, atomic_amd64x.go is just one of them.
```

I wonder if anyone has run into this issue, and my question would be i see in registry-image-resource it's unpacking layers in reverse order, it's ok if the existing files gets modified in later layers but i dont' see it handling older layer files getting removed in later layers ? maybe someone can educate me on this a bit.

Thanks

**Comment by chenbh** [comment]
This is indeed a bug with the registry resource, I've submitted an issue at https://github.com/concourse/registry-image-resource/issues/267. 

In the meantime, it should work if you split the go installation command into 2:
```dockerfile
RUN rm -rf /usr/local/go && tar -C /usr/local -xzf go1.16.linux-amd64.tar.gz
```

To:
```dockerfile
RUN rm -rf /usr/local/go
RUN tar -C /usr/local -xzf go1.16.linux-amd64.tar.gz
```
